File: cli/commands/cache.py
from __future__ import annotations

import copy
from enum import Enum
from typing import Optional

import typer
from rich.console import Console
from rich.table import Table

from etfagents.default_config import DEFAULT_CONFIG
from etfagents.cache_manager import CacheManager

_console = Console()

cache_app = typer.Typer(help="Cache management utilities.")


class _CacheTypeAll(str, Enum):
    # cleanup uses Optional[_CacheTypeAll] (None=all default); clear requires explicit choice including "all"
    api = "api"
    signals = "signals"
    snapshots = "snapshots"
    checkpoints = "checkpoints"
    all = "all"


_CacheType = _CacheTypeAll


@cache_app.command("stats")
def cache_stats(
    json_output: bool = typer.Option(False, "--json", help="Output as JSON."),
) -> None:
    config = copy.deepcopy(DEFAULT_CONFIG)
    mgr = CacheManager(config)
    s = mgr.stats()

    if json_output:
        _console.print_json(data=s)
        return

    table = Table(title="Cache Statistics", show_footer=True)
    table.add_column("Category", style="cyan")
    table.add_column("Files", justify="right", style="green")
    table.add_column("Size (MB)", justify="right", style="yellow")

    total_files = 0
    for cat in CacheManager.CATEGORIES:
        info = s[cat]
        total_files += info["count"]
        table.add_row(cat, str(info["count"]), f"{info['size_mb']:.2f}")

    table.add_row("[bold]Total[/bold]", f"[bold]{total_files}[/bold]", f"[bold]{s['total_mb']:.2f}[/bold]")
    _console.print(table)


@cache_app.command("cleanup")
def cache_cleanup(
    days: int = typer.Option(7, "--days", min=0, help="Remove entries older than N days. 0=clear all."),
    category: Optional[_CacheType] = typer.Option(None, "--type", help="Cache category. Default: all."),
) -> None:
    config = copy.deepcopy(DEFAULT_CONFIG)
    mgr = CacheManager(config)
    cat = category.value if category else "all"
    result = mgr.cleanup(days, cat)
    _console.print(f"Deleted {result['deleted_files']} file(s), freed {result['freed_mb']:.2f} MB")


@cache_app.command("clear")
def cache_clear(
    category: _CacheTypeAll = typer.Option(_CacheTypeAll.all, "--type", help="Cache category to clear."),
    confirm: bool = typer.Option(False, "--yes", "-y", help="Skip confirmation prompt."),
) -> None:
    if not confirm:
        if category.value == "all":
            message = "Clear ALL cache categories? This cannot be undone"
        else:
            message = f"Clear all '{category.value}' cache? This cannot be undone"
        confirmed = typer.confirm(message)
        if not confirmed:
            raise typer.Exit(code=0)

    config = copy.deepcopy(DEFAULT_CONFIG)
    mgr = CacheManager(config)
    result = mgr.clear(category.value)
    _console.print(f"Cleared {result['deleted_files']} file(s), freed {result['freed_mb']:.2f} MB")

File: ts/src/cli/commands/cache.ts
/**
 * `cache <action>` — manage bridge cache.
 *
 * Phase 3.5: wraps bridge RPCs cache.stats / cache.cleanup / cache.clear.
 */

import type { Command } from "commander";
import pc from "picocolors";
import { BridgeApi, BridgeClient, RpcError } from "../../bridge/index.js";

export function registerCache(program: Command): void {
  const cacheCmd = program.command("cache").description("Manage bridge cache");

  cacheCmd
    .command("stats")
    .description("Show cache statistics")
    .action(async () => {
      const client = new BridgeClient();
      const api = new BridgeApi(client);
      try {
        await client.start();
        const stats = await api.cacheStats();
        console.log(JSON.stringify(stats, null, 2));
      } catch (err) {
        handleError(err);
      } finally {
        await client.close();
      }
    });

  cacheCmd
    .command("cleanup")
    .description("Remove stale cache entries")
    .option("--days <n>", "Remove entries older than N days", "30")
    .action(async (opts: { days: string }) => {
      const client = new BridgeClient();
      const api = new BridgeApi(client);
      try {
        await client.start();
        const result = await api.cacheCleanup(Number(opts.days));
        console.log(result);
      } catch (err) {
        handleError(err);
      } finally {
        await client.close();
      }
    });

  cacheCmd
    .command("clear")
    .description("Clear all cached data")
    .option("--type <type>", "Cache type to clear", "all")
    .option("--yes", "Skip confirmation")
    .action(async (opts: { type: string; yes?: boolean }) => {
      if (!opts.yes) {
        console.log(
          pc.yellow(`This will clear all '${opts.type}' cache entries. Use --yes to confirm.`),
        );
        return;
      }
      const client = new BridgeClient();
      const api = new BridgeApi(client);
      try {
        await client.start();
        const result = await api.cacheClear(opts.type as "all");
        console.log(result);
      } catch (err) {
        handleError(err);
      } finally {
        await client.close();
      }
    });
}

function handleError(err: unknown): void {
  if (err instanceof RpcError) {
    console.error(pc.red(`bridge error [${err.code}]: ${err.message}`));
  } else {
    console.error(pc.red(`error: ${(err as Error).message}`));
  }
  process.exitCode = 1;
}
